src/openrct2/drawing/X8DrawingEngine.cpp

```diff
@@ -403,24 +403,34 @@ void X8DrawingEngine::DrawAllDirtyBlocks()
             uint32_t columns = xx - x;
 
             // Check rows
-            uint32_t yy;
-            for (yy = y; yy < dirtyBlockRows; yy++)
+            uint32_t yy = y;
+            if (CheckRows(yy, y, xx, x, dirtyBlockRows, dirtyBlockColumns, columns, dirtyBlocks))
             {
-                uint32_t yyOffset = yy * dirtyBlockColumns;
-                for (xx = x; xx < x + columns; xx++)
-                {
-                    if (dirtyBlocks[yyOffset + xx] == 0)
-                    {
-                        goto endRowCheck;
-                    }
-                }
+                uint32_t rows = yy - y;
+                DrawDirtyBlocks(x, y, columns, rows);
             }
+        }
+    }
+}
 
-        endRowCheck:
-            uint32_t rows = yy - y;
-            DrawDirtyBlocks(x, y, columns, rows);
+bool X8DrawingEngine::CheckRows(
```

Comment:
Rename the function GetNumDirtyRows. Then you could make it return `yy-y` and avoid the pass by reference discussed in the above. Would work something like this:

```
...::DrawAllDirtyBlocks{
.... L403
auto rows = GetNumDirtyRows(x, y, columns);
DrawDirtyBlocks(x, y, columns, rows);
... L411
}
uint32_t ...::GetNumDirtyRows(const uint32_t x, const uint32_t y, const uint32_t columns)
{
    const uint32_t dirtyBlockColumns = _dirtyGrid.BlockColumns;
    const uint32_t dirtyBlockRows = _dirtyGrid.BlockRows;
    const uint8_t* dirtyBlocks = _dirtyGrid.Blocks;
...
for...
    for...
      if ...
         return yy - y;
 ...
 return yy - y;
}

Reply:
I'm sorry i don't get it: GetNumDirtyBlocks needs both xx and yy to function but they are not passed as arguments, also, if i move dirtyBlockColumns, dirtyBlockRows, and dirtyBlocks* inside the body of GetNumDirtyRows then DrawAllDirtyBlocks would not be able to run it's for loops in the beginning of the function. I'm probably missing something.

Reply:
Xx is just a temporary variable for the for loop as is yy but for the outer loop. As for the dirty block variables I'm not saying move them I'm saying have a duplicate of them in this function.